api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java

```diff
@@ -0,0 +1,285 @@
+import org.openmrs.module.htmlformentry.widget.Widget;
+
+public class ConditionElement implements HtmlGeneratorElement, FormSubmissionControllerAction  {
+	
+	private MessageSourceService mss;
+	private ConditionService conditionService;
+	private boolean required;
+	// widgets
+	private Widget conditionName;
+	private DateWidget onSetDate;
+	private DateWidget endDate;
+	private RadioButtonsWidget conditionStatusWidget;
+	private ErrorWidget endDateErrorWidget;
+	private ErrorWidget conditionNameErrorWidget;
+	private ErrorWidget conditionStatusErrorWidget;
+	
+	@Override
+	public void handleSubmission(FormEntrySession session, HttpServletRequest submission) {
+		FormEntryContext context = session.getContext();
+		if (context.getMode() != Mode.VIEW) {
+			Condition condition = new Condition();
+			CodedOrFreeText conditionConcept = new CodedOrFreeText();
+			try {
+				int conceptId = Integer.parseInt((String) conditionName.getValue(session.getContext(), submission));
+				conditionConcept.setCoded(new Concept(conceptId));
+				
+			} catch(NumberFormatException e) {
+				String nonCodedConcept = submission.getParameter(context.getFieldName(conditionName));
+				conditionConcept.setNonCoded(nonCodedConcept);
+			}
+			condition.setCondition(conditionConcept);
+			condition.setClinicalStatus(getStatus(context, submission));
+			condition.setOnsetDate(onSetDate.getValue(context, submission));
+			if (ConditionClinicalStatus.INACTIVE == getStatus(context, submission)) {
+				condition.setEndDate(endDate.getValue(context, submission));
+			}
+			condition.setPatient(session.getPatient());
+			conditionService = Context.getConditionService();
+			conditionService.saveCondition(condition);
+		}
+	}
+
+	@Override
+	public Collection<FormSubmissionError> validateSubmission(FormEntryContext context, HttpServletRequest submission) {		
+		List<FormSubmissionError> ret = new ArrayList<FormSubmissionError>();
+		Date givenOnsetDate = onSetDate.getValue(context, submission);
+		Date givenEndDate = endDate.getValue(context, submission);
+		String condition = submission.getParameter(context.getFieldName(conditionName));
+		ConditionClinicalStatus status = getStatus(context, submission);
+		
+		if (context.getMode() != Mode.VIEW) {
+			if (StringUtils.isBlank(condition) && required) {
+				ret.add(new FormSubmissionError(context
+	                    .getFieldName(conditionName), "htmlformentryui.conditionui.condition.required"));
+			}	
+			if (givenOnsetDate != null && givenEndDate != null) {
+				if (givenOnsetDate.after(givenEndDate)) {
+					ret.add(new FormSubmissionError(context
+	                        .getFieldName(endDate), "htmlformentryui.conditionui.endDate.before.onsetDate.error"));
+				}
+			} 
+			if (status == null && (required || StringUtils.isNotBlank(condition))) {
+				ret.add(new FormSubmissionError(context
+	                    .getFieldName(conditionStatusWidget), "htmlformentryui.conditionui.status.required"));
+			}
```

Comment:
I'm confused now, I thought all those validations were ensured at the service layer level?

Reply:
I think this is required before the form is submitted!

- It prevents the form from being submitted once a condition is required and one isn't given ie. `<condition required="true" />` .
- It prevents form submission when also one provided the Condition and didn't select a status. This applies to cases where the condition is option ie. `<condition required="false"/>`

Reply:
The case of `required` should be handled here for sure, i.e. making sure that the condition field is filled in if the tag is required.

>It prevents form submission when also one provided the Condition and didn't select a status.

But that case should be picked up by the service layer no?

Reply:
Ok. I guess have fixed that?